src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractChangeNamespaceService.cs

```diff
@@ -0,0 +1,573 @@
+
+            // Here's the entire process for changing namespace:
+            // 1. Change the namespace declaration, fix references and add imports that might be necessary.
+            // 2. Explicitly merge the diff to get a new solution.
+            // 3. Remove added imports that are unnecessary.
+            // 4. Do another explicit diff merge based on last merged solution.
+            //
+            // The reason for doing explicit diff merge twice is so merging after remove unnecessaty imports can be correctly handled.
+
+            var solutionAfterNamespaceChange = solution;
+            var referenceDocuments = PooledHashSet<DocumentId>.GetInstance();
+
+            foreach (var id in documentIds)
+            {
+                var (newSolution, refDocumentIds) = await ChangeNamespaceToMatchFoldersAsync(solutionAfterNamespaceChange, id, declaredNamespace, targetNamespace, cancellationToken).ConfigureAwait(false);
+                solutionAfterNamespaceChange = newSolution;
+                referenceDocuments.AddRange(refDocumentIds);
+            }
+
+            var solutionAfterFirstMerge = await MergeDiffAsync(solution, solutionAfterNamespaceChange, cancellationToken).ConfigureAwait(false);
+
+            // After changing documents, we still need to remove unnecessary imports related to our change.
+            // We don't try to remove all imports that might become unnecessary/invalid after the namespace change, 
+            // just ones that fully matche the old/new namespace. Because it's hard to get it right and will almost 
+            // certainly cause perf issue.
+            // For example, if we are changing namespace `Foo.Bar` (which is the only namespace declaration with such name)
+            // to `A.B`, the using of name `Bar` in a different file below would remain untouched, even it's no longer valid:
+            //
+            //      namespace Foo
+            //      {
+            //          using Bar;
+            //          ~~~~~~~~~
+            //      }
+            //
+            // Also, because we may have added different imports to document that triggered the refactoring
+            // and the documents that reference affected types declared in changed namespace, we try to remove
+            // unnecessary imports separately.
+
+            var solutionAfterImportsRemoved = await RemoveUnnecessaryImportsAsync(
+                solutionAfterFirstMerge,
+                documentIds,
+                CreateAllContainingNamespaces(declaredNamespace),
+                cancellationToken).ConfigureAwait(false);
+
+            solutionAfterImportsRemoved = await RemoveUnnecessaryImportsAsync(
+                solutionAfterImportsRemoved,
+                referenceDocuments.ToImmutableArray(),
+                ImmutableArray.Create(declaredNamespace, targetNamespace),
+                cancellationToken).ConfigureAwait(false);
+
+            referenceDocuments.Free();
```

Comment:
minor, but it's nice to generally have 'free's in finallies.  THat way we clean up even when cancelling.  But this is very very minor, especially if this is during code action invocatin, and not during analysis.